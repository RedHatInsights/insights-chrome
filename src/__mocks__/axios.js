import mockAxios from 'jest-mock-axios';
module.exports = mockAxios;
