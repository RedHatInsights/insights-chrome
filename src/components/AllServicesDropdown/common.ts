import { createContext } from 'react';

export const TAB_CONTENT_ID = 'refTab1Section';
export const FAVORITE_TAB_ID = 'favorites';

export const AllServicesDropdownContext = createContext<{ onLinkClick: () => void }>({
  onLinkClick: () => undefined,
});
