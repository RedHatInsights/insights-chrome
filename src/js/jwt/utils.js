const pub = {};

pub.exposeTest = (priv) => {
    if (window.NODE_ENV === 'test') {
        module.exports.priv = priv;
    }
};

module.exports = pub;
