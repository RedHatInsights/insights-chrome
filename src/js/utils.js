import get from 'lodash/get';

function getWindow() {
    return window;
}

/* eslint-disable curly */
export function isValidAccountNumber(num) {
    if (!num) return false;
    if (num === -1) return false;
    if (num === '-1') return false;
    return Number.isInteger(Number(num));
}
/* eslint-enable curly */

function getSection() {
    const sections = getWindow().location.pathname.split('/');
    if (sections[1] === 'beta') { return sections[2] || ''; }

    return sections[1];
}

export function pageAllowsUnentitled() {
    const pathname = getWindow().location.pathname;
    if (pathname === '/' ||
        pathname === '/beta' ||
        pathname === '/beta/' ||
        pathname.indexOf('/openshift') === 0 ||
        pathname.indexOf('/beta/openshift') === 0) {
        return true;
    }

    return false;
}

export function pageRequiresAuthentication() {
    const section = getSection();
    if (section === 'insights' ||
        section === 'rhel'     ||
        section === 'hybrid'   ||
        section === 'apps'     ||
        section === 'ansible'  ||
        section === 'migrations') {
        return true;
    }

    return false;
}

/**
 * Creates a redux listener that watches the state on given path (e.g. chrome.appNav) and calls
 * the given function when the state on the given path changes.
 *
 * The function is called with two parameters: current state value on the path, store reference
 */
export function createReduxListener(store, path, fn) {
    let previous = undefined;

    return () => {
        const state = store.getState();
        const current = get(state, path);

        if (current !== previous) {
            previous = current;
            fn(current, store);
        }
    };
}
