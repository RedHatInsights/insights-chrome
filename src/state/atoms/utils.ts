import { WritableAtom, atom } from 'jotai';

// recipe from https://jotai.org/docs/recipes/atom-with-toggle
export function atomWithToggle(initialValue?: boolean, onToggle?: (value: boolean) => void): WritableAtom<boolean, [boolean?], void> {
  const anAtom = atom(initialValue, (get, set, nextValue?: boolean) => {
    const update = nextValue ?? !get(anAtom);
    set(anAtom, update);
    onToggle?.(update);
  });

  return anAtom as WritableAtom<boolean, [boolean?], void>;
}
