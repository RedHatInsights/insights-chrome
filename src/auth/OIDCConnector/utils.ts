import { AuthContextProps } from 'react-oidc-context';
import { ITLess, LOGIN_SCOPES_STORAGE_KEY, deleteLocalStorageItems } from '../../utils/common';
import { GLOBAL_FILTER_KEY, OFFLINE_REDIRECT_STORAGE_KEY } from '../../utils/consts';
import Cookies from 'js-cookie';
import logger from '../logger';
import createUUID from './createUUID';

const log = logger('auth:utils');

enum AllowedPartnerScopes {
  aws = 'aws',
  azure = 'azure',
  gcp = 'gcp',
}

function isPartnerScope(scope: string): scope is AllowedPartnerScopes {
  return Object.values(AllowedPartnerScopes).includes(scope as AllowedPartnerScopes);
}

function getPartnerScope(pathname: string) {
  // replace beta and leading "/"
  const sanitizedPathname = pathname.replace(/^(\/beta\/|\/preview\/)/, '/').replace(/^\//, '');
  // check if the pathname is connect/:partner
  if (sanitizedPathname.match(/^connect\/.+/)) {
    // return :partner param
    const fragmentScope = sanitizedPathname.split('/')[1];
    if (isPartnerScope(fragmentScope)) {
      return `api.partner_link.${fragmentScope}`;
    }
    log(`Invalid stratosphere scope: ${fragmentScope}`);
    return undefined;
  }

  return undefined;
}

export async function logout(auth: AuthContextProps, bounce?: boolean) {
  const keys = Object.keys(localStorage).filter(
    (key) =>
      key.endsWith('/api/entitlements/v1/services') ||
      key.endsWith('/chrome') ||
      key.endsWith('/chrome-store') ||
      key.startsWith('kc-callback') ||
      key.startsWith(GLOBAL_FILTER_KEY)
  );
  deleteLocalStorageItems([...keys, OFFLINE_REDIRECT_STORAGE_KEY, LOGIN_SCOPES_STORAGE_KEY]);
  if (bounce) {
    const eightSeconds = new Date(new Date().getTime() + 8 * 1000);
    Cookies.set('cs_loggedOut', 'true', {
      expires: eightSeconds,
    });
    await auth.signoutRedirect({
      redirectTarget: 'top',
      post_logout_redirect_uri: window.location.origin,
      id_token_hint: undefined,
    });
  } else {
    await auth.revokeTokens(['access_token', 'refresh_token']);
  }
}

export function login(auth: AuthContextProps, requiredScopes: string[] = [], redirectUri = location.href) {
  log('Logging in');
  // Redirect to login
  Cookies.set('cs_loggedOut', 'false');
  //FIX ME: Temp fix until scope is added in-boundary
  let scope = ITLess() ? ['openid', ...requiredScopes] : ['openid', ...requiredScopes];
  const partner = getPartnerScope(window.location.pathname);
  if (partner) {
    scope.push(partner);
  }
  scope = Array.from(new Set(scope));
  localStorage.setItem(LOGIN_SCOPES_STORAGE_KEY, JSON.stringify(scope));
  // KC scopes are delimited by a space character, hence the join(' ')
  return auth.signinRedirect({
    redirect_uri: redirectUri,
    scope: scope.join(' '),
    nonce: createUUID(),
  });
}
