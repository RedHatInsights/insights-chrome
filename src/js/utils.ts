import get from 'lodash/get';
import { Store } from 'redux';
import { DEFAULT_ROUTES } from './jwt/constants';
import flatMap from 'lodash/flatMap';
import { NavItem } from './types';
import axios from 'axios';
import { ChromeModule, RouteDefinition } from './redux/store';

export function getWindow() {
  return window;
}

export function isValidAccountNumber(num?: number | string) {
  if (!num) return false;
  if (num === -1) return false;
  if (num === '-1') return false;
  return Number.isInteger(Number(num));
}

export function getSection() {
  const sections = getWindow().location.pathname.split('/');
  if (sections[1] === 'beta') {
    return sections[2] || '';
  }

  return sections[1];
}

export function pageAllowsUnentitled() {
  const pathname = getWindow().location.pathname;
  if (
    pathname === '/' ||
    pathname === '/beta' ||
    pathname === '/beta/' ||
    pathname.indexOf('/openshift') === 0 ||
    pathname.indexOf('/beta/openshift') === 0 ||
    pathname.indexOf('/security') === 0 ||
    pathname.indexOf('/beta/security') === 0 ||
    pathname.indexOf('/application-services') === 0 ||
    pathname.indexOf('/beta/application-services') === 0 ||
    pathname.indexOf('/hac') === 0 ||
    pathname.indexOf('/beta/hac') === 0 ||
    pathname.indexOf('/ansible/ansible-dashboard/trial') === 0 ||
    pathname.indexOf('/beta/ansible/ansible-dashboard/trial') === 0
  ) {
    return true;
  }

  return false;
}

export function pageRequiresAuthentication() {
  const section = getSection();
  if (
    section === 'insights' ||
    section === 'cost-management' ||
    section === 'apps' ||
    section === 'ansible' ||
    section === 'migrations' ||
    section === 'subscriptions' ||
    section === 'openshift' ||
    section === 'settings' ||
    section === 'user-preferences' ||
    section === 'internal' ||
    section === 'application-services'
  ) {
    return true;
  }

  return false;
}

/**
 * Creates a redux listener that watches the state on given path (e.g. chrome.appNav) and calls
 * the given function when the state on the given path changes.
 *
 * The function is called with two parameters: current state value on the path, store reference
 */
export function createReduxListener(store: Store, path: string, fn: (current: any, store: Store) => void) {
  let previous: any = undefined;

  return () => {
    const state = store.getState();
    const current = get(state, path);

    if (current !== previous) {
      previous = current;
      fn(current, store);
    }
  };
}

export function deleteLocalStorageItems(keys: string[]) {
  keys.map((key) => localStorage.removeItem(key));
}

export function lastActive(searchString: string, fallback: string) {
  return Object.keys(localStorage).reduce<string | { expires: string }>((acc, curr) => {
    if (curr.includes(searchString)) {
      try {
        let accDate;
        try {
          const localStorageDate = localStorage.getItem(acc as string);
          accDate = localStorageDate ? new Date(JSON.parse(localStorageDate)?.expires) : new Date();
        } catch {
          accDate = new Date();
        }

        const currObj = JSON.parse(localStorage.getItem(curr) || '');
        return accDate >= new Date(currObj.expires) ? acc : curr;
      } catch (e) {
        return acc;
      }
    }

    return acc;
  }, fallback);
}

export const isAnsible = (sections: string[]) => (sections.includes('ansible') && sections.includes('insights') ? 1 : 0);

export function getUrl(type?: string) {
  if (window.location.pathname === '/beta/' || window.location.pathname === '/') {
    return 'landing';
  }

  const sections = window.location.pathname.split('/');
  if (sections[1] === 'beta') {
    return type === 'bundle' ? sections[2] : sections[3 + isAnsible(sections)];
  }

  return type === 'bundle' ? sections[1] : sections[2 + isAnsible(sections)];
}

export function getEnv() {
  return Object.entries(DEFAULT_ROUTES).find(([, { url }]) => url.includes(location.hostname))?.[0] || 'qa';
}

export function getEnvDetails() {
  return Object.entries(DEFAULT_ROUTES).find(([, { url }]) => url.includes(location.hostname))?.[1];
}

export function isProd() {
  return location.host === 'cloud.redhat.com' || location.host === 'console.redhat.com' || location.host.includes('prod.foo.redhat.com');
}

export function isBeta() {
  return window.location.pathname.split('/')[1] === 'beta' ? true : false;
}

export function isFedRamp() {
  return getEnv() === 'gov';
}

export function updateDocumentTitle(title?: string, noSuffix = false) {
  const titleSuffix = '| console.redhat.com';
  if (typeof title === 'undefined') {
    return;
  }
  if (typeof title === 'string') {
    document.title = title.includes(titleSuffix) || noSuffix ? title : `${title} ${titleSuffix}`;
  } else {
    console.warn(`Title is not a string. Got ${typeof title} instead.`);
  }
}

const activateChild = (hrefMatch: string, childRoutes: NavItem[]) => {
  let hasActiveChild = false;
  const routes = childRoutes.map((item) => {
    const active = item.href === hrefMatch;
    if (active) {
      hasActiveChild = true;
    }
    return {
      ...item,
      active,
    };
  });
  return {
    active: hasActiveChild,
    routes,
  };
};

function mutateSchema(hrefMatch: string, navItems: NavItem[]): NavItem[] {
  return navItems.map((item) => {
    const { href, routes, navItems } = item;
    if (!href && navItems) {
      return {
        ...item,
        navItems: mutateSchema(hrefMatch, navItems),
      };
    }

    if (!href && routes) {
      return {
        ...item,
        ...activateChild(hrefMatch, routes),
      };
    }

    if (href) {
      return {
        ...item,
        active: item.href === hrefMatch,
      };
    }

    return item;
  });
}

export const highlightItems = (pathname: string, navItems: NavItem[], sortedLinks: string[]) => {
  const cleanPathname = pathname.replace(/\/$/, '');
  const segmentsCount = cleanPathname.split('/').length + 1;
  const matchedLink = sortedLinks.find((href) => {
    const segmentedHref = href.replace(/\/$/, '').split('/').slice(0, segmentsCount).join('/');
    return cleanPathname.includes(segmentedHref);
  });
  return mutateSchema(matchedLink?.replace(/\/$/, '') || '', navItems);
};

export const levelArray = (navItems: NavItem[]): string[] => {
  return flatMap<NavItem, string>(navItems, ({ href, routes, navItems }) => {
    if (!href && navItems) {
      return levelArray(navItems);
    }

    if (!href && routes) {
      return levelArray(routes);
    }

    if (href) {
      return [href];
    }

    return [];
  });
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function noop() {}

export const trustarcScriptSetup = () => {
  const trustarcScript = document.createElement('script');
  trustarcScript.id = 'trustarc';

  if (location.host === 'console.redhat.com') {
    trustarcScript.src = '//static.redhat.com/libs/redhat/marketing/latest/trustarc/trustarc.js';
  } else {
    trustarcScript.src = '//static.redhat.com/libs/redhat/marketing/latest/trustarc/trustarc.stage.js';
  }

  document.body.appendChild(trustarcScript);
};

export const loadFedModules = () =>
  axios.get(`${window.location.origin}${isBeta() ? '/beta' : ''}/config/chrome/fed-modules.json?ts=${Date.now()}`, {
    headers: {
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      Expires: '0',
    },
  });

export const generateRoutesList = (modules: { [key: string]: ChromeModule }) =>
  Object.entries(modules)
    .reduce<RouteDefinition[]>(
      (acc, [scope, { dynamic, manifestLocation, isFedramp, modules = [] }]) => [
        ...acc,
        ...modules
          .map(({ module, routes }) =>
            /**Clean up this map function */
            routes.map((route) => ({
              scope,
              module,
              isFedramp: typeof route === 'string' ? isFedramp : route.isFedramp,
              path: typeof route === 'string' ? route : route.pathname,
              manifestLocation,
              dynamic: typeof dynamic === 'boolean' ? dynamic : typeof route === 'string' ? true : route.dynamic,
              exact: typeof route === 'string' ? false : route.exact,
            }))
          )
          .flat(),
      ],
      []
    )
    .sort((a, b) => (a.path.length < b.path.length ? 1 : -1));
