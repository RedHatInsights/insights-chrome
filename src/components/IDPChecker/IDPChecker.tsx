import React, { Fragment, useEffect, useRef, useState } from 'react';
import axios from 'axios';
import { useSelector } from 'react-redux';

import { ITLess } from '../../utils/common';
import IDPError from '../ErrorComponents/IDPError';
import { ReduxState } from '../../redux/store';

const IDPStatuses = {
  OK: 'OK',
  UNKNOWN: 'UNKNOWN',
  PENDING: 'PENDING',
  ERROR: 'ERROR',
};

const IDPChecker: React.FunctionComponent = ({ children }) => {
  const ITLessEnv = ITLess();
  const missingIDP = useSelector(({ chrome }: ReduxState) => chrome?.missingIDP);
  const [status, setStatus] = useState(() => {
    if (ITLessEnv) {
      return missingIDP === true ? IDPStatuses.ERROR : IDPStatuses.UNKNOWN;
    }
    return IDPStatuses.OK;
  });
  const hasUser = useSelector(({ chrome: { user } }: ReduxState) => Object.keys(user || {}).length > 0);
  const allowStateChange = useRef(ITLessEnv);

  useEffect(() => {
    if (ITLessEnv && status !== IDPStatuses.PENDING && hasUser) {
      allowStateChange.current && setStatus(IDPStatuses.PENDING);
      axios
        .get('/api/entitlements/v1/services')
        .then(() => {
          allowStateChange.current && setStatus(IDPStatuses.OK);
        })
        .catch((err) => {
          const authError = err.response.status === 403 && err.message === 'Insights authorization failed - account number not in allow list';
          allowStateChange.current && setStatus(authError ? IDPStatuses.ERROR : IDPStatuses.OK);
        });
    }
  }, [hasUser, missingIDP]);

  useEffect(() => {
    if (missingIDP === true) {
      allowStateChange.current && setStatus(IDPStatuses.ERROR);
      allowStateChange.current = false;
    }
  }, [missingIDP]);

  if (status === IDPStatuses.OK) {
    return <Fragment>{children}</Fragment>;
  }
  if (status === IDPStatuses.ERROR) {
    return <IDPError />;
  }
  return null;
};

export default IDPChecker;
