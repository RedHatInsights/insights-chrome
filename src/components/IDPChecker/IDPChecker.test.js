import React from 'react';
import axios from 'axios';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { act } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';
import * as utils from '../../utils/common';
import IDPChecker from './IDPChecker';

jest.mock('../../utils/common', () => {
  const utils = jest.requireActual('../../utils/common');
  return {
    __esModule: true,
    ...utils,
    ITLess: jest.fn(),
  };
});

jest.mock('axios', () => {
  const axios = jest.requireActual('axios');
  return {
    __esModule: true,
    ...axios,
    get: jest.fn(),
  };
});

describe('<IDPChecker />', () => {
  const ITLessSpy = jest.spyOn(utils, 'ITLess');
  const getSpy = jest.spyOn(axios, 'get');
  let mockStore;
  const initialState = {
    chrome: {
      user: {
        foo: 'bar',
      },
      missingIDP: false,
    },
  };

  beforeEach(() => {
    mockStore = configureStore();
    ITLessSpy.mockReturnValue(true);
    getSpy.mockImplementation(() => Promise.resolve({}));
  });

  test('should render children in non fedRamp env', () => {
    ITLessSpy.mockReturnValueOnce(false);
    const store = mockStore(initialState);
    const { container, queryAllByTestId } = render(
      <Provider store={store}>
        <IDPChecker>
          <span data-testid="foo">OK</span>
        </IDPChecker>
      </Provider>
    );

    expect(queryAllByTestId('foo')).toHaveLength(1);
    expect(container).toMatchSnapshot();
  });

  test('should render error state if IDP is missing in fedramp env', async () => {
    const store = mockStore({
      ...initialState,
      chrome: {
        missingIDP: true,
      },
    });

    await act(async () => {
      render(
        <Provider store={store}>
          <IDPChecker>
            <span data-testid="foo">OK</span>
          </IDPChecker>
        </Provider>
      );
    });

    expect(screen.queryAllByTestId('foo')).toHaveLength(0);
    expect(screen.getAllByText('Authorization failure')).toHaveLength(1);
  });

  test('should render error state if IDP test API returns 403', async () => {
    getSpy.mockImplementationOnce(() =>
      Promise.reject({
        response: {
          status: 403,
        },
        message: 'Insights authorization failed - account number not in allow list',
      })
    );
    const store = mockStore(initialState);

    await act(async () => {
      render(
        <Provider store={store}>
          <IDPChecker>
            <span data-testid="foo">OK</span>
          </IDPChecker>
        </Provider>
      );
    });

    expect(screen.queryAllByTestId('foo')).toHaveLength(0);
    expect(screen.getAllByText('Authorization failure')).toHaveLength(1);
  });

  test('should render children if IDP validation passes', async () => {
    const store = mockStore(initialState);

    await act(async () => {
      render(
        <Provider store={store}>
          <IDPChecker>
            <span data-testid="foo">OK</span>
          </IDPChecker>
        </Provider>
      );
    });

    expect(screen.queryAllByTestId('foo')).toHaveLength(1);
  });
});
