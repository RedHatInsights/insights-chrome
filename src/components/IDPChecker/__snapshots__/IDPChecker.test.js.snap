// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IDPChecker /> should render children in non fedRamp env 1`] = `
<div>
  <span
    data-testid="foo"
  >
    OK
  </span>
</div>
`;
